### PR TITLE
[incubator-kie-issues#1541] Add “evaluationHitIds” to the response of jitexecutor-dmn’s “/dmnresult” route

### LINKED ISSUE #1541: Add “evaluationHitIds” to the response of jitexecutor-dmn’s “/dmnresult” route
Modify jitexecutor-dmn| so that the execution rest endpoint (/jitdmn/dmnresult) returns information about executed branches of the DMN model.

The format of returned json should be (e.g.)
```json
{
    "namespace": "https://kie.org/dmn/_A3317FB1-7BF8-4904-A5F4-2CD63AF3AEC9",
    "modelName": "DMN_A77074C1-21FE-4F7E-9753-F84661569AFC",
    "dmnContext": {
        "Risk Score": 20,
        "DTI": 15,
        "Loan Pre-Qualification": "Qualified",
        "id": "_CD8A1707-4F28-4F99-9D07-340F8E1A82AC",
        "Credit Score": "Excellent"
    },
    "messages": [],
    "decisionResults": [
        {
            "decisionId": "_6C207FD6-14EA-4480-A90D-074ABC97B54B",
            "decisionName": "Risk Score",
            "result": 20,
            "messages": [],

"evaluationHitIds": [<rule_id1>, <rule_id2>],

            "evaluationStatus": "SUCCEEDED"
        },
        {
            "decisionId": "_A1236581-4334-43F1-82A0-72CE204F07F9",
            "decisionName": "Loan Pre-Qualification",
            "result": "Qualified",
            "messages": [],
            "evaluationHitIds": [<expression_id1>],
            "evaluationStatus": "SUCCEEDED"
        }
    ]
} 
```
where `<rule_id1>, <rule_id2>`  referes to id of rows executed inside decision tables, and `<expression_id1>` refers to branches (then/else) of `conditional` element

e.g. 
```xml
<rule id="_1578BD9E-2BF9-4BFC-8956-1A736959C937">
        <inputEntry id="_57DAA79A-4929-4D0C-A7D1-E7BA81674B8B">
          <text>&quot;Poor&quot;</text>
        </inputEntry>
        <inputEntry id="_E2D67EAF-9023-4657-A852-7AFF89DDFF26">
          <text>-</text>
        </inputEntry>
        <outputEntry id="_93601F57-F06C-4AD2-9BA0-6933AD8F6007">
          <text>30</text>
        </outputEntry>
        <annotationEntry>
          <text>// Your annotations here</text>
        </annotationEntry>
      </rule>
```

returned id: "_1578BD9E-2BF9-4BFC-8956-1A736959C937"

```xml
<conditional id="_A80832EF-8F9C-4638-B5FB-047210860DF2" label="Loan Pre-Qualification" typeRef="Loan Pre-Qualification">
      <if id="_3C702CE4-E5A0-4B6F-905D-C2621FFFA387">
        <literalExpression id="_CA44F978-C3A3-4D65-BABB-8BDA2FBB1074" label="Expression Name">
          <text>Risk Score &lt;= 30</text>
        </literalExpression>
      </if>
      <then id="_6481FF12-61B5-451C-B775-4143D9B6CD6B">
        <literalExpression id="_000EE7D4-6DAB-4BC1-8A5A-59E1E9E61B21" label="Expression Name">
          <text>&quot;Qualified&quot; </text>
        </literalExpression>
      </then>
      <else id="_2CD02CB2-6B56-45C4-B461-405E89D45633">
        <literalExpression id="_EDEFF23B-55B9-4D8C-AB04-8E49F642CB22" label="Expression Name">
          <text>&quot;Not Qualified&quot;</text>
        </literalExpression>
      </else>
    </conditional>
```

returned id: "_6481FF12-61B5-451C-B775-4143D9B6CD6B" (_then_) OR "_2CD02CB2-6B56-45C4-B461-405E89D45633" (_else_)

### PR DESCRIPTION
**Pull request by gitgabrio**
Fixes https://github.com/apache/incubator-kie-issues/issues/1541

This PR:

1. add `Set<String> evaluationHitIds` field to `JITDMNResult`
2. define a specific `JITDMNListener` that store evaluated ids
3. add the above listener to `DMNRuntime` used behind `/dmnresult` rest endpoint
4. add the retrieved ids to the returned `JITDMNResult`

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] tests</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] quarkus-lts</b>
 
- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] native</b>

- for <b>native lts checks</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins run native-lts</b>

- for a <b>specific native lts check</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] native-lts</b>
 
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

<details>
<summary>
Quarkus-3 PR check is failing ... what to do ?
</summary>
The Quarkus 3 check is applying patches from the `.ci/environments/quarkus-3/patches`.

The first patch, called `0001_before_sh.patch`, is generated from Openrewrite `.ci/environments/quarkus-3/quarkus3.yml` recipe. The patch is created to speed up the check. But it may be that some changes in the PR broke this patch.  
No panic, there is an easy way to regenerate it. You just need to comment on the PR:
```
jenkins rewrite quarkus-3
```
and it should, after some minutes (~20/30min) apply a commit on the PR with the patch regenerated.

Other patches were generated manually. If any of it fails, you will need to manually update it... and push your changes.
</details>

### CODE REVIEW

File: jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/JITDMNListener.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
```
Comment (yesamer) [inline jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/JITDMNListener.java]:
```suggestion
 *
```

File: jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/JITDMNListener.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
```
Comment (yesamer) [inline jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/JITDMNListener.java]:
```suggestion
 *
```

File: jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluator.java [hunk 3]
```diff
@@ -99,7 +110,8 @@ public static DMNEvaluator fromMultiple(MultipleResourcesPayload payload) {
             }
         }
         if (mainModel == null) {
-            throw new IllegalStateException("Was not able to identify main model from MultipleResourcesPayload contents.");
+            throw new IllegalStateException("Was not able to identify main model from MultipleResourcesPayload " +
+                    "contents.");
```
Comment (yesamer) [inline jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluator.java]: @gitgabrio autoformatting? :D

**Review by yesamer — APPROVED** [review]
Thank you @gitgabrio, your changes are fine. Just some autoformatting points

File: jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluator.java [hunk 4]
```diff
@@ -22,7 +22,10 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
```
Comment (jomarko) [inline jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluator.java]: this is maybe unused now?

File: jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluator.java [hunk 5]
```diff
@@ -73,9 +78,16 @@ public Collection<DMNModel> getAllDMNModels() {
...
+    public JITDMNResult evaluate(Map<String, Object> context) {
+        DMNContext dmnContext =
+                new DynamicDMNContextBuilder(dmnRuntime.newContext(), dmnModel).populateContextWith(context);
+        DMNResult dmnResult = dmnRuntime.evaluateAll(dmnModel, dmnContext);
+        Optional<List<String>> evaluationHitIds = dmnRuntime.getListeners().stream()
+                .filter(JITDMNListener.class::isInstance)
+                .findFirst()
+                .map(JITDMNListener.class::cast)
+                .map(JITDMNListener::getEvaluationHitIds);
+        return new JITDMNResult(getNamespace(), getName(), dmnResult, evaluationHitIds.orElse(Collections.emptySet()));
```
Comment (jomarko) [inline jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluator.java]: `emptyList`

**Review by jomarko — COMMENTED** [review]
@gitgabrio I put some comments to code, I didn't checked added tests for now, will do later, probably once we figure out if it is possible to hit single rule multiple times in single evaluation. See the private discussion with @ljmotta

File: jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluator.java (hunk 5 above)
Comment (gitgabrio) [reply jitexecutor/jitexecutor-dmn/src/main/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluator.java]: Thanks for spotting!

**Comment by kie-ci3** [comment]
**PR job** `#4` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-apps -u https://github.com/apache/incubator-kie-kogito-apps/pull/2120 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/4/display/redirect

**Test results:**
- PASSED: 1648
- FAILED: 2

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/4/testReport/org.kie.kogito.it.jobs/WorkflowTimeoutsIT/PR_check___Build_projects___workflowTimeoutExceeded/">PR check / Build projects / org.kie.kogito.it.jobs.WorkflowTimeoutsIT.workflowTimeoutExceeded</a></summary>
java.lang.IllegalStateException: Unable to determine the status of the running process. See the above logs for details
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/4/testReport/org.kie.kogito.jobs.service.resource/KeycloakPostgreSqlJobResourceTest/___/">org.kie.kogito.jobs.service.resource.KeycloakPostgreSqlJobResourceTest.(?)</a></summary>
java.lang.RuntimeException: Unable to start Quarkus test resource class org.kie.kogito.testcontainers.quarkus.PostgreSqlQuarkusTestResource
</details>

**Review by jomarko — APPROVED** [review]
Thank you @gitgabrio

**Review by baldimir — APPROVED** [review]

**Comment by gitgabrio** [comment]
kogito-apps failure unrelated

> Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.8.0:dockerBuild (default) on project integration-tests-trusty-service-springboot: Unknown mediaType: application/vnd.oci.image.index.v1+json

**Comment by kie-ci3** [comment]
**PR job** `#7` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-apps -u https://github.com/apache/incubator-kie-kogito-apps/pull/2120 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/7/display/redirect

**Test results:**
- PASSED: 1662
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/7/testReport/org.kie.kogito.explainability.messaging/ExplainabilityMessagingHandlerCounterfactualsIT/explainabilityRequestIsProcessedAndAnIntermediateMessageIsSent/">org.kie.kogito.explainability.messaging.ExplainabilityMessagingHandlerCounterfactualsIT.explainabilityRequestIsProcessedAndAnIntermediateMessageIsSent</a></summary>
<br/>Wanted but not invoked:<br/>explanationServiceImpl.explainAsync(<br/>    <any org.kie.kogito.explainability.api.BaseExplainabilityRequest>,<br/>    <any><br/>);<br/>-> at org.kie.kogito.explainability.messaging.BaseExplainabilityMessagingHandlerIT.explainabilityRequestIsProcessedAndAnIntermediateMessageIsSent(BaseExplainabilityMessagingHandlerIT.java:140)<br/>Actually, there were zero interactions with this mock.<br/>
</details>

**Comment by kie-ci3** [comment]
**PR job** `#9` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-apps -u https://github.com/apache/incubator-kie-kogito-apps/pull/2120 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/9/display/redirect

**Test results:**
- PASSED: 1677
- FAILED: 7

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/9/testReport/org.kie.kogito.index.postgresql/ProcessDataIndexPostgreSqlHttpIT/testProcessInstanceEvents/">org.kie.kogito.index.postgresql.ProcessDataIndexPostgreSqlHttpIT.testProcessInstanceEvents</a></summary>
java.util.concurrent.CompletionException: java.lang.RuntimeException: Unable to start Quarkus test resource class org.kie.kogito.index.test.quarkus.http.DataIndexPostgreSqlHttpQuarkusTestResource
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/9/testReport/org.kie.kogito.index.postgresql/ProcessDataIndexPostgreSqlKafkaIT/testProcessInstanceEvents/">org.kie.kogito.index.postgresql.ProcessDataIndexPostgreSqlKafkaIT.testProcessInstanceEvents</a></summary>
java.util.concurrent.CompletionException: java.lang.RuntimeException: Unable to start Quarkus test resource class org.kie.kogito.index.test.quarkus.kafka.DataIndexPostgreSqlQuarkusKafkaTestResource
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/9/testReport/org.kie.kogito.index/ProcessDataIndexMongoDBIT/testProcessInstanceEvents/">org.kie.kogito.index.ProcessDataIndexMongoDBIT.testProcessInstanceEvents</a></summary>
Cannot invoke method size() on null object
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/9/testReport/org.kie.kogito.index/ProcessDataIndexPostgreSqlIT/testProcessInstanceEvents/">org.kie.kogito.index.ProcessDataIndexPostgreSqlIT.testProcessInstanceEvents</a></summary>
Failed to load ApplicationContext for [WebMergedContextConfiguration@50e3019b testClass = org.kie.kogito.index.ProcessDataIndexPostgreSqlIT, locations = [], classes = [org.kie.kogito.KogitoApplication], contextInitializerClasses = [org.kie.kogito.index.spring.KogitoServiceRandomPortSpringTestResource, org.kie.kogito.index.spring.DataIndexPostgreSqlSpringTestResource], activeProfiles = [], propertySourceDescriptors = [], propertySourceProperties = ["org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true", "server.port=0"], contextCustomizers = [org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@6f1c29b7, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@2a7ed1f, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5456afaa, org.springframework.boot.test.autoconfigure.actuate.observability.ObservabilityContextCustomizerFactory$DisableObservabilityContextCustomizer@1f, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizer@1787bc24, org.springframework.boot.test.context.SpringBootTestAnnotation@bf04646b], resourceBasePath = "src/main/webapp", contextLoader = org.springframework.boot.test.context.SpringBootContextLoader, parent = null]
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/9/testReport/org.kie.kogito.it.jobs/ProcessTimerIT/PR_check___Build_projects___testDeleteTimerCycle/">PR check / Build projects / org.kie.kogito.it.jobs.ProcessTimerIT.testDeleteTimerCycle</a></summary>
java.util.concurrent.CompletionException: java.lang.RuntimeException: Unable to start Quarkus test resource class org.kie.kogito.test.resources.JobServiceCompositeQuarkusTestResource
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/9/testReport/org.kie.kogito.addons.quarkus.data.index.it/PostgreSQLQuarkusAddonDataIndexPersistenceIT/PR_check___Build_projects___testDataIndexAddon/">PR check / Build projects / org.kie.kogito.addons.quarkus.data.index.it.PostgreSQLQuarkusAddonDataIndexPersistenceIT.testDataIndexAddon</a></summary>
java.util.concurrent.CompletionException: java.lang.RuntimeException: Unable to start Quarkus test resource class org.kie.kogito.index.test.quarkus.http.DataIndexPostgreSqlHttpQuarkusTestResource
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2120/9/testReport/org.kie.kogito.addons.quarkus.data.index.it/PostgreSQLQuarkusAddonDataIndexPersistenceIT/PR_check___Build_projects___testDataIndexAddon/">PR check / Build projects / org.kie.kogito.addons.quarkus.data.index.it.PostgreSQLQuarkusAddonDataIndexPersistenceIT.testDataIndexAddon</a></summary>
java.lang.RuntimeException: Unable to start Quarkus test resource class org.kie.kogito.index.test.quarkus.http.DataIndexPostgreSqlHttpQuarkusTestResource
</details>

**Comment by gitgabrio** [comment]
jenkins failure unrelated

> Caused by: com.google.api.client.http.HttpResponseException: 502 Bad Gateway

**Pull request merged**